### PR TITLE
Make public asset use explicit

### PR DESCRIPTION
**Pull request by schneems**
Currently when an asset isn't found the behavior is to pass the string through. For example a valid asset will return a url from pipeline

``` ruby
asset_path("application.js")
# => assets/application-123098udasvi0mnafd.js
```

While if you make a typo, you won't get an error or anything, it just falls through:

``` ruby
asset_path("app1icati0n.js")
# => "app1icati0n.js"
```

This is bad. No exceptions are raised, the error goes undetected.

This PR makes adds an explicit api for when you want to use an asset that is not in the asset pipeline. If you want an asset from the "public" directory you can now explicitly say so by using `public_folder: true`

``` ruby
asset_path("app1icati0n.js", public_folder: true)
# => "app1icati0n.js"
```

Now there is no confusion by sprockets-rails about the intent of the method call. Likewise in rails/sprockets-rails#375 we are adding a flag `unknown_asset_fallback`. When it is set to `true` then the current fallback behavior will happen, but a deprecation will be emitted. When it is set to false an exception will be raised so you know immediately that the asset was not found.

We are defaulting it to `false` for all new apps.

``` ruby
# Unknown asset fallback will return the path passed in when the given
# asset is not present in the asset pipeline.
Rails.application.config.assets.unknown_asset_fallback = false
```

While this patch is intended to go along with the sprockets-rails patch, it should still function with older sprockets-rails versions, however no deprecations or exceptions will be raised.

### CODE REVIEW

**Comment by schneems** [comment]
The sprockets-rails PR is https://github.com/rails/sprockets-rails/pull/375

File: actionview/lib/action_view/helpers/asset_tag_helper.rb [hunk 1]
```diff
@@ -304,19 +344,27 @@ def audio_tag(*sources)
         multiple_sources_tag("audio", sources)
       end
 
+      # Returns an HTML audio tag for the source asset in the public
+      # folder. This uses +audio_tag+ and skips any asset
```
Comment (alexcameron89) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: You skipped this line in the documentation here: `# lookups by assuming any assets are in the `public` folder.`

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 2]
```diff
@@ -158,6 +167,14 @@ def asset_path(source, options = {})
       end
       alias_method :path_to_asset, :asset_path # aliased to avoid conflicts with an asset_path named route
 
+      # Computes the path to an asset in the public folder.
+      # This uses +asset_path+ and skips any asset lookups by assuming the asset is in the
```
Comment (alexcameron89) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: On these docs, should it be `+path_to_asset+`? Same question for the ones below.

**Comment by matthewd** [comment]
This feels too onerous to me. Particularly, having the asset pipeline hijack `image_tag` seems out of step with how other similar tag helpers behave.

**Comment by schneems** [comment]
@md not sure I follow. Asset pipeline already hijacks these helpers. This patch provides an un-hijack.

**Comment by robin850** [comment]
:+1: on deprecating but can't we just add the `:public_folder` option to the existing helpers and output the warning if none of `:public_folder` or `:raw` is passed ? Thus we don't have to maintain more methods and more documentation for two things that are mostly the same.

**Comment by matthewd** [comment]
@schneems the distinction I'm making is that it currently *enhances* the helper, but doesn't interfere with non-asset use.

This also creates a cascade problem, where because `image_path` becomes `public_image_path`, `image_tag` must become `public_image_tag`, and so must some 3rd party helper that calls `image_tag`, etc.

I propose that we add `public_image_path`, and deprecate calling `image_path` with a **relative** filename that is not a known asset... and that's all.

So, anyone using `image_tag` with an asset, or a full image URL, or `/some/thing.png`, is unaffected -- all of those strings mean exactly one thing, and there's no potential for misinterpretation. Anyone currently using `image_tag('my/image.png')` to refer to something in `public/images/my/image.png` will be warned to change that to: `image_tag(public_image_path('my/image.png'))` -- I feel no obligation to offer a `public_image_tag`, because that formulation is relatively rare. And so the 3rd party helper will also be unaffected: it takes a string in, and the application-level caller introduces the `public_image_path` helper if they're doing the unusual thing.

**Comment by schneems** [comment]
I never liked the idea of introducing so many new helpers, method lookup time isn't free and more methods can mean slower performance over time. I do like the idea of just exposing the `public_folder: true` API. It's basically almost already implemented by this patch, and makes intent clear.

The code in rails needs to enable people to run the same app they do today without deprecations. It also needs to be written in such a way that we can detect correctly when we should be deprecating (and eventually raising). I'm not in love with the idea of something like `image_tag(public_image_path('my/image.png'))`, it looks a bit gnarly and the only way I can see to detect when it's not being used correctly is to have `public_image_path` return a special object instead of a string.

I'm going to explore modifying this patch to get rid of the `public_*` helpers, which I think everyone agrees are not the best idea.

## Side Notes

I wanted to note that I don't like the API of passing in an asset that starts with a slash as indicating a full path. I deprecated that. I think it's too easy to accidentally start an asset with a slash and not realize that you're bypassing the whole asset pipeline, silently. At the end of the day I want an error to show up in development if someone uses one of these helpers with anything that won't resolve to an asset that exists and render in production.

I also wanted to note a special case in `asset_path` for the full url option: When you call `image_tag` with something like `image_tag("https://www.schneems.com/smile.png")` the call will eventually call `asset_path` before it calls `compute_asset_path` which is the main entry point to the asset pipeline. Before that can happen

```ruby
def asset_path(source, options = {})
  raise ArgumentError, "nil is not a valid asset source" if source.nil?

  source = source.to_s
  return "" if source.blank?
  return source if URI_REGEXP.match?(source) # <====== HERE
```

This last line will return true since we're giving it a full URI so the asset lookup never happens, no deprecation is thrown. So we don't need to handle that case for skipping deprecations.

Thanks for all then input. Let me know if you need more clarification on any of this, or want to respond to any of the sub-points.

**Comment by matthewd** [comment]
> I wanted to note that I don't like the API of passing in an asset that starts with a slash as indicating a full path. I deprecated that.

IMO it's no more acceptable for `image_tag('/a/url')` to start insisting on assets than it would be for `link_to('/a/url')` to start complaining about the argument not matching a route: they're both direct convenience wrappers for their corresponding HTML tags, and it's only a secondary / additional convenience that they can intuit a "better" bonus behaviour for some arguments.

> I think it's too easy to accidentally start an asset with a slash and not realize that you're bypassing the whole asset pipeline, silently.

It's not supposed to be; that's why we turned on digests in development. If you use `/assets/my/image.png`, that's already supposed to complain & fail. And if you use `/my/image.png`, it won't complain, but it also won't ever work -- and I could likely be convinced that should raise **if** `my/image.png` is a valid asset.

The only other possibility is that someone manually determines and then uses `/assets/my/image-012345etc.png`. That's going to go pretty poorly, but feels like a reasonable limit-point to me... the gun is braced to point horizontally and down-range: if you still manage to get your foot in front of it, good luck to you.

(Which of those three is the one that worries you? The first certainly used to be a big problem, but that's why we fixed it... or at least I thought we did.)

---

> I'm not in love with the idea of something like image_tag(public_image_path('my/image.png')), it looks a bit gnarly 

My first thought was to just kill the `/images` magic completely... given that the asset pipeline has been the "right" path for a long time now, I'm dubious that it sees much use. And people that *are* using it could just switch to `/images/my/image.png` -- the magic of the asset pipeline seems worthwhile for what it gives you... inserting 8 characters feels a bit more opaque-for-the-sake-of-it.

With that context, the above doesn't trouble me overmuch... the two methods are at very different conceptual levels, just like `link_to(new_post_comment_path(@post))`. Nice API is nice, but we're not here to prevent people from ever having to call more than one method.

> and the only way I can see to detect when it's not being used correctly is to have public_image_path return a special object instead of a string.

It'd be returning an absolute URL `/images/my/image.png`, whereas I only want to complain about a bare `my/image.png`.

> This last line will return true since we're giving it a full URI so the asset lookup never happens, no deprecation is thrown.

:+1:

**Comment by schneems** [comment]
I updated to use `public_folder: true` as the API and got rid of the `public_*` methods. Need to update sprockets-rails patch.

**Comment by schneems** [comment]
Updated sprockets-rails patch, was pretty minimal.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/assets.rb.tt [hunk 3]
```diff
@@ -6,6 +6,10 @@ Rails.application.config.assets.version = '1.0'
 # Add additional assets to the asset load path
 # Rails.application.config.assets.paths << Emoji.images_path
 
+# Unknown asset fallback will return the path passed in when the given
+# asset is not present in the asset pipeline.
+Rails.application.config.assets.unknown_asset_fallback = false
```
Comment (matthewd) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/assets.rb.tt]: I think this has to go elsewhere so the upgrade task can skip it :confused:

Comment (schneems) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/assets.rb.tt]: Moved to the new initializer

File: guides/source/asset_pipeline.md [hunk 4]
```diff
@@ -586,6 +586,19 @@ in your application are included in the `config.assets.precompile` list.
 If `config.assets.digest` is also true, the asset pipeline will require that
 all requests for assets include digests.
 
+### Raise an Error When an Asset is Not Found
+
+If you are using a recent version of sprockets-rails you can configure what happens
```
Comment (prathamesh-sonpatki) [inline guides/source/asset_pipeline.md]: may be specify exactly which version of sprockets-rails, like greater than > 3.something ?

**Comment by schneems** [comment]
Updated to `Rails.application.config.assets.unknown_asset_fallback = <%= options[:update] ? true : false %>` and mentioned the specific version of sprockets-rails in the assets guide,

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 5]
```diff
@@ -118,16 +118,67 @@ module Helpers
     module AssetUrlHelper
       URI_REGEXP = %r{^[-a-z]+://|^(?:cid|data):|^//}i
 
-      # Computes the path to asset in public directory. If :type
-      # options is set, a file extension will be appended and scoped
-      # to the corresponding public directory.
+      # This is the entry point for all assets.
+      # When using the asset pipeline (i.e. sprockets and sprockets-rails), the
+      # behavior is "enhanced". You can bypass the asset pipeline by passing in
+      # `public_folder: true` to the options.
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: The RDoc parser doesn't understand back-ticks, you should rather use `<tt>public_folder: true</tt>`.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 6]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
-      #   asset_path "application", type: :stylesheet     # => /assets/application.css
-      #   asset_path "http://www.example.com/js/xmlhr.js" # => http://www.example.com/js/xmlhr.js
+      # === With the asset pipeline
+      #
+      # All options passed to asset_path will be passed to `compute_asset_path
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (`public_folder: true`)
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto, use `<tt>` here.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 7]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
       #
       # All other asset *_path helpers delegate through this method.
       #
-      #   asset_path "application.js"                     # => /assets/application.js
-      #   asset_path "application", type: :javascript     # => /assets/application.js
-      #   asset_path "application", type: :stylesheet     # => /assets/application.css
-      #   asset_path "http://www.example.com/js/xmlhr.js" # => http://www.example.com/js/xmlhr.js
+      # === With the asset pipeline
+      #
+      # All options passed to asset_path will be passed to `compute_asset_path
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Here you could write `+asset_path+` and `+compute_asset_path+`; if you don't wrap `asset_path` around pluses, RDoc will generate a link back to this method, which is pretty useless.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 8]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (`public_folder: true`)
+      #
+      # Accepts a `type` option that can specify the asset's extension. No error
+      # checking is done to verify the source passed into `asset_path` is valid
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto, you can use pluses here.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 9]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      # === With the asset pipeline
+      #
+      # All options passed to asset_path will be passed to `compute_asset_path
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (`public_folder: true`)
+      #
+      # Accepts a `type` option that can specify the asset's extension. No error
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 10]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      # and that the file exists on disk.
+      #
+      #      asset_path("application.js", public_folder: true)                 # => "application.js"
+      #      asset_path("filedoesnotexist.png", public_folder: true)           # => "filedoesnotexist.png"
+      #      asset_path("application", type: :javascript, public_folder: true) # => "/javascripts/application.js"
+      #      asset_path("application", type: :stylesheet, public_folder: true) # => "/stylesheets/application.css"
+      #
+      # === Options applying to all assets
+      #
+      # Below lists scenarios that apply to `asset_path` whether or not you're
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 11]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      #   urls and will not be expanded. This will bypass the asset pipeline.
+      #
+      #     asset_path("/foo.png") # => "/foo.png"
+      #
+      # - All blank strings will be returned immediately. This bypasses the
+      #   asset pipeline and all other behavior described.
+      #
+      #     asset_path("") # => ""
+      #
+      # - If `config.relative_url_root` is specified, all assets will have that
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 12]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      #   root prepended.
+      #
+      #     Rails.application.config.relative_url_root = "bar"
+      #     asset_path("foo.js", public_folder: true) # => "bar/foo.js"
+      #
+      # - A different asset host can be specified via `config.action_controller.asset_host`
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 13]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      #   this is commonly used in conjunction with a CDN.
+      #
+      #     Rails.application.config.action_controller.asset_host = "assets.example.com"
+      #     asset_path("foo.js", public_folder: true) # => "http://assets.example.com/foo.js"
+      #
+      # - An extension name can be specified manually with `extname`.
```
Comment (robin850) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ditto.

File: guides/source/asset_pipeline.md [hunk 14]
```diff
@@ -724,7 +737,7 @@ If you have other manifests or individual stylesheets and JavaScript files to
 include, you can add them to the `precompile` array in `config/initializers/assets.rb`:
 
 ```ruby
-Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += ['admin.js', 'admin.css', 'swfObject.js']
```
Comment (robin850) [inline guides/source/asset_pipeline.md]: Really nit-picky but why did you change the style here ? This is not consistent with other code snippets and the code generated by Rails by default. :blush:

File: guides/source/configuring.md [hunk 15]
```diff
@@ -169,6 +169,8 @@ pipeline is enabled. It is set to `true` by default.
 
 * `config.assets.precompile` allows you to specify additional assets (other than `application.css` and `application.js`) which are to be precompiled when `rake assets:precompile` is run.
 
+* `config.assets.unknown_asset_fallback` allows you to modify the behavior of the asset pipeline when an asset is not in the pipeline with recent verisons of sprockets-rails.
```
Comment (robin850) [inline guides/source/configuring.md]: Also nit-picky but to be consistent, you could say like above, "in the pipeline, if you use sprockets-rails 3.2.0 or newer".

**Comment by robin850** [comment]
This is looking very good ! :+1:

Just a nit-pick again regarding documentation though :smile: it looks like you didn't document the `:public_poster_folder` option for the `video_tag` method, this may be useful.

File: guides/source/asset_pipeline.md [hunk 16]
```diff
@@ -586,6 +586,19 @@ in your application are included in the `config.assets.precompile` list.
...
+when an asset lookup is performed and nothing is found. If you turn off "asset fallback"
+then an error will be raised when an asset cannot be found.
+
+```ruby
+config.assets.unknown_asset_fallback = false
+```
+
+If "asset fallback" is enabled then when an asset cannot be found the path will be
+output instead and no error raised.
+
```
Comment (vipulnsward) [inline guides/source/asset_pipeline.md]: We should mention what the default behaviour is here.

File: guides/source/configuring.md [hunk 17]
```diff
@@ -169,6 +169,8 @@ pipeline is enabled. It is set to `true` by default.
 
 * `config.assets.precompile` allows you to specify additional assets (other than `application.css` and `application.js`) which are to be precompiled when `rake assets:precompile` is run.
 
+* `config.assets.unknown_asset_fallback` allows you to modify the behavior of the asset pipeline when an asset is not in the pipeline, if you use sprockets-rails 3.2.0 or newer.
```
Comment (vipulnsward) [inline guides/source/configuring.md]: Ditto, we should mention, `defaults to ...`

**Comment by vipulnsward** [comment]
@schneems is there a way to test the request *does* send in requested public resource, I see the test cases for debug output that check something does not exist.

**Comment by schneems** [comment]
@vipulnsward there is one test with `application.js` with and without `public_folder: true`. Check out `test "{ public_folder: true } does not use the asset pipeline" do`

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 18]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      # All options passed to +asset_path+ will be passed to +compute_asset_path+
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (<tt>public_folder: true</tt>)
+      #
+      # Accepts a <tt>type</tt> option that can specify the asset's extension. No error
+      # checking is done to verify the source passed into +asset_path+ is valid
+      # and that the file exists on disk.
```
Comment (dhh) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Not the biggest fan of `public_folder`. It implies that we're doing something special with this, when all we do is pass this directly through. I'd think something negative like `skip_pipeline: true` would be clearer for that. Would match this documentation as well.

File: actionview/lib/action_view/helpers/asset_tag_helper.rb [hunk 19]
```diff
@@ -269,6 +270,8 @@ def image_alt(src)
       #   # => <video preload="none" controls="controls" src="/videos/trailer.ogg" ></video>
       #   video_tag("trailer.m4v", size: "16x10", poster: "screenshot.png")
       #   # => <video src="/videos/trailer.m4v" width="16" height="10" poster="/assets/screenshot.png"></video>
+      #   video_tag("trailer.m4v", size: "16x10", poster: "screenshot.png", public_poster_folder: true)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: Shouldn't this just be `skip_pipeline` to match the other options? Presumably the same reasons renaming `public_folder` to `skip_pipeline` works here, so what do we gain making this stand out?

Comment (schneems) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: There need to be 2 options here as the video may be in the pipeline but the poster might not be. Good catch though on the naming. I'll fix this.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 20]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
-      #   asset_path "application.js"                     # => /assets/application.js
-      #   asset_path "application", type: :javascript     # => /assets/application.js
-      #   asset_path "application", type: :stylesheet     # => /assets/application.css
-      #   asset_path "http://www.example.com/js/xmlhr.js" # => http://www.example.com/js/xmlhr.js
+      # === With the asset pipeline
+      #
+      # All options passed to +asset_path+ will be passed to +compute_asset_path+
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: I think our style is to nest code examples by 3 spaces, not 6.

File: actionview/lib/action_view/helpers/asset_url_helper.rb [hunk 21]
```diff
@@ -118,16 +118,67 @@ module Helpers
...
+      # - A different asset host can be specified via <tt>config.action_controller.asset_host</tt>
+      #   this is commonly used in conjunction with a CDN.
+      #
+      #     Rails.application.config.action_controller.asset_host = "assets.example.com"
+      #     asset_path("foo.js", skip_pipeline: true) # => "http://assets.example.com/foo.js"
+      #
+      # - An extension name can be specified manually with <tt>extname</tt>.
+      #
+      #     asset_path("foo", skip_pipeline: true, extname: ".js")     # => "/foo.js"
+      #     asset_path("foo.css", skip_pipeline: true, extname: ".js") # => "/foo.css.js"
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Should these last 3 examples use `skip_pipeline`?

File: railties/test/application/asset_debugging_test.rb [hunk 22]
```diff
@@ -68,5 +68,102 @@ class ::PostsController < ActionController::Base ; end
...
+        image_tag:              %r{<img src="/images/#{ contents }"},
+        favicon_link_tag:       %r{<link rel="shortcut icon" type="image/x-icon" href="/images/#{ contents }" />},
+        stylesheet_link_tag:    %r{<link rel="stylesheet" media="screen" href="/stylesheets/#{ contents }.css" />},
+        javascript_include_tag: %r{<script src="/javascripts/#{ contents }.js">},
+        audio_tag:              %r{<audio src="/audios/#{ contents }"></audio>},
+        video_tag:              %r{<video src="/videos/#{ contents }"></video>}
+      }
+
+      cases.each do |(view_method, tag_match)|
+        app_file "app/views/posts/index.html.erb", "<%= #{ view_method } '#{contents}', skip_pipeline: true %>"
```
Comment (kaspth) [inline railties/test/application/asset_debugging_test.rb]: Mixed spaces around interpolation braces abound in this test file. I think our preferred style is to not have any around interpolation braces, e.g.: `#{view_method}`.

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 21 above)
Comment (schneems) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Makes the example more clear IMHO

File: actionview/lib/action_view/helpers/asset_tag_helper.rb [hunk 23]
```diff
@@ -282,8 +285,11 @@ def image_alt(src)
       #   video_tag(["trailer.ogg", "trailer.flv"], size: "160x120")
       #   # => <video height="120" width="160"><source src="/videos/trailer.ogg" /><source src="/videos/trailer.flv" /></video>
       def video_tag(*sources)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: We could remove the options wrangling with a keyword argument:

```ruby
def video_tag(*sources, skip_pipeline_on_poster: nil)
  multiple_sources_tag_builder("video", sources) do |options|
    options[:poster] = path_to_image(options[:poster], skip_pipeline: skip_pipeline_on_poster) if options[:poster]
    options[:width], options[:height] = extract_dimensions(options.delete(:size)) if options[:size]
  end
end
```

I think that would also save us a block argument shadowing local variable warning from `options`. 😁

File: actionview/lib/action_view/helpers/asset_tag_helper.rb (hunk 19 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: Ah gotcha!

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 21 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Yet I was confused by the usage 😄

I'm not objecting to using the option necessarily, but why does using it make the examples more clear?

Comment (schneems) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: The assetpipeline does a bunch of things and I don't want asset pipeline magic being confused with this option. Adding a long fingerprint to the example makes it less clear that an extension was added to the output. What it doesn't do is replace an extension name, so `foo.css` does not become `foo.js`. Also technically this option doesn't change the extension output, it changes the input that goes into the asset pipeline. It's harder to show.

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 20 above)
Comment (schneems) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: When you do that the example lines up with the bullet points, is that correct?

```
      # - An extension name can be specified manually with <tt>extname</tt>.
      #
      #   asset_path("foo", skip_pipeline: true, extname: ".js")     # => "/foo.js"
      #   asset_path("foo.css", skip_pipeline: true, extname: ".js") # => "/foo.css.js"
```

File: actionview/lib/action_view/helpers/asset_tag_helper.rb (hunk 23 above)
Comment (schneems) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: It's turtles all the way down, other method signatures that this gets passed to use extract_options.

Better to do that in a separate PR.

Comment (schneems) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: Oh, i see what you're saying. I still want to do that in a different PR.

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 20 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ah, overlooked the bullet points! I guess the style then is 2 spaces from the starting indentation of a sentence. So here it would be 5 spaces.

Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: In fact the examples seem to mix 6 and 5 spaces, so let's just have them all be 5 😁

File: actionview/lib/action_view/helpers/asset_tag_helper.rb (hunk 23 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb]: I'm not generally objecting to `extract_options!`, I agree that's out of scope. I'm objecting to using it in `video_tag` because the diff says it didn't use it before. But messing around in IRB showed me that we can't mix keyword arguments with a symbol keyed options hash without using `**options`, so that doesn't save us much here anyway. 👍 to keeping it like you had it 😁

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 21 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Okay, sounds good 👍

File: actionview/lib/action_view/helpers/asset_url_helper.rb (hunk 20 above)
Comment (schneems) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Some examples are bullet points and some aren't which is why they are different. Thanks for the reviews btw.

Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Then they should only have had 3 space indentation 😉

Comment (kaspth) [inline actionview/lib/action_view/helpers/asset_url_helper.rb]: Ah sorry, they did get shortened to 3 spaces! GitHub somehow showed me an older diff.
